// frontend/src/lib/api.ts
'use server';

import { Software, Category, Criterion } from './types';

const SOFTWARE_DB: Software[] = [
    {
        id: 1,
        name: 'Next.js',
        slug: 'nextjs',
        shortDescription: 'The React Framework for the Web',
        categoryIds: [1],
        logoUrl: 'https://www.vectorlogo.zone/logos/nextjs/nextjs-icon.svg',
        features: { 1: 4.5, 2: 4.2, 3: 4.8, 4: 5, 6: true },
    },
    {
        id: 2,
        name: 'Nest.js',
        slug: 'nestjs',
        shortDescription: 'A progressive Node.js framework',
        categoryIds: [1],
        logoUrl: 'https://www.vectorlogo.zone/logos/nestjs/nestjs-icon.svg',
        features: { 1: 4.3, 2: 3.8, 3: 4.5, 4: 4.2, 6: true },
    },
    {
        id: 3,
        name: 'Tailwind CSS',
        slug: 'tailwindcss',
        shortDescription: 'Utility-first CSS framework',
        categoryIds: [3],
        logoUrl: 'https://www.vectorlogo.zone/logos/tailwindcss/tailwindcss-icon.svg',
        features: { 1: 4.8, 2: 4.7, 3: 4.9, 4: 4.8, 6: true },
    },
    {
        id: 4,
        name: 'PostgreSQL',
        slug: 'postgresql',
        developer: 'Postgres',
        shortDescription: 'Advanced open source relational database',
        fullDescription: '## Overview\nPostgreSQL is a powerful, open source relational database management system (RDBMS).',
        categoryIds: [2, 1, 3, 4],
        logoUrl: 'https://www.vectorlogo.zone/logos/postgresql/postgresql-icon.svg',
        websiteUrl: 'https://www.postgresql.org/',
        githubUrl: 'https://github.com/postgres/postgres',
        screenshots: ['https://www.vectorlogo.zone/logos/dotnet/dotnet-official.svg', 'https://www.vectorlogo.zone/logos/dotnet/dotnet-official.svg', 'https://www.vectorlogo.zone/logos/postgresql/postgresql-ar21~bgwhite.svg', 'https://www.vectorlogo.zone/logos/postgresql/postgresql-ar21~bgwhite.svg', 'https://www.vectorlogo.zone/logos/postgresql/postgresql-ar21~bgwhite.svg'],
        features: { 1: 4.9, 2: 3.5, 3: 4.6, 4: 4.8, 6: true },
    },
    {
        id: 5,
        name: 'TypeScript TEXT TEXT',
        slug: 'typescript',
        shortDescription:
            'Statically typed, interpreted, and compiled high-level language',
        categoryIds: [4],
        logoUrl: 'https://www.vectorlogo.zone/logos/typescriptlang/typescriptlang-icon.svg',
        features: { 1: 4.6, 2: 3.9, 3: 4.7, 4: 4.9, 6: true },
    },
    {
        id: 6,
        name: 'Express.js',
        slug: 'expressjs',
        shortDescription:
            'Fast, unopinionated, minimalist web framework for Node.js',
        categoryIds: [1],
        features: { 1: 4.2, 2: 4.5, 3: 3.8, 4: 4.6, 6: true },
    },
    {
        id: 7,
        name: 'React',
        slug: 'react',
        shortDescription: 'JavaScript library for building user interfaces',
        categoryIds: [1],
        features: { 1: 4.7, 2: 4.1, 3: 4.3, 4: 5, 6: true },
    },
    {
        id: 8,
        name: 'Node.js',
        slug: 'nodejs',
        shortDescription:
            "JavaScript runtime built on Chrome's V8 JavaScript engine",
        categoryIds: [1],
        features: { 1: 4.4, 2: 4.0, 3: 4.1, 4: 4.7, 6: true },
    },
];

const CATEGORY_DB: Category[] = [
    { id: 1, name: 'Frameworks', slug: 'frameworks' },
    { id: 2, name: 'Databases', slug: 'databases' },
    { id: 3, name: 'CSS Tools', slug: 'css-tools' },
    { id: 4, name: 'Languages', slug: 'languages' },
];

export async function searchAction(query: string): Promise<Software[]> {
    await new Promise((resolve) => setTimeout(resolve, 300));

    if (!query || query.trim().length < 2) return [];

    const q = query.toLowerCase();

    const softwareResults = SOFTWARE_DB.filter(
        (item) =>
            item.name.toLowerCase().includes(q) ||
            item.shortDescription.toLowerCase().includes(q),
    );

    return [...softwareResults];
}

export async function getSoftwareBySlug(slug: string): Promise<Software | null> {
    await new Promise((resolve) => setTimeout(resolve, 100));
    return SOFTWARE_DB.find(item => item.slug === slug) || null;
}

export async function getFeaturedAction(): Promise<Software[]> {
    await new Promise((resolve) => setTimeout(resolve, 100));

    const randomIndex = Math.floor(Math.random() * SOFTWARE_DB.length);
    const startIndex = Math.max(0, randomIndex - 2);
    const endIndex = Math.min(SOFTWARE_DB.length, startIndex + 3);

    return SOFTWARE_DB.slice(startIndex, endIndex);
}
export async function getSoftwareByIdAction(id: number): Promise<Software | undefined> {
    await new Promise((resolve) => setTimeout(resolve, 100));

    return SOFTWARE_DB.find((item) => item.id === id);
}
export async function getSoftwareBySlugAction(slug: string): Promise<Software | undefined> {
    await new Promise((resolve) => setTimeout(resolve, 100));

    return SOFTWARE_DB.find((item) => item.slug === slug);
}
export async function getCategoryByIdAction(id: number): Promise<Category | undefined> {
    await new Promise((resolve) => setTimeout(resolve, 100));

    return CATEGORY_DB.find((item) => item.id === id);
}

const CRITERIA_DB: Criterion[] = [
    { id: 1, name: 'Performance', type: 'rating', weight: 5 },
    { id: 2, name: 'Ease of Use', type: 'rating', weight: 4 },
    { id: 3, name: 'Documentation', type: 'rating', weight: 3 },
    { id: 4, name: 'Community Support', type: 'rating', weight: 4 },
    { id: 5, name: 'Price', type: 'currency', weight: 3 },
    { id: 6, name: 'Open Source', type: 'boolean', weight: 2 },
];

export async function getCriteriaAction(): Promise<Criterion[]> {
    await new Promise((resolve) => setTimeout(resolve, 100));

    return CRITERIA_DB;
}
